File: csharp/exercise/exer05a-product-matrix-vector.cs
/*
 * MATRIX-VECTOR MULTIPLICATION
 */
using System;
using System.Collections.Generic;
using System.Threading;



class Exer05A : IRunnable
{
    public void run()
    {
        double[][] A = new double[][] {
            new double[] { 1, 2, 3 },
            new double[] { 4, 5, 6 },
            new double[] { 7, 8, 9 }
        };

        double[] b = {
            3,
            -1,
            0
        };

        var result = getProduct(A, b);

        Array.ForEach(result, Console.WriteLine);
    }


    private double[] getProduct(double[][] mat, double[] vec)
    {
        // Assume that size of mat and vec are both correct
        int sizeRowMat = mat.Length;
        // int sizeColMat = mat[0].Length;
        // int sizeVec = vec.Length;

        var result = new double[sizeRowMat];
        var lstTh = new List<Thread>();

        for (int ith = 0; ith < sizeRowMat; ++ith)
        {
            int i = ith;

            lstTh.Add(new Thread(() =>
            {
                var u = mat[i];
                var v = vec;
                result[i] = Exer05Util.getScalarProduct(u, v);
            }));
        }

        lstTh.ForEach(th => th.Start());
        lstTh.ForEach(th => th.Join());

        return result;
    }
}

File: cpp/cpp-boost/exer05a-product-matrix-vector.cpp
/*
MATRIX-VECTOR MULTIPLICATION
*/


#include <iostream>
#include <vector>
#include <boost/assign/std/vector.hpp>
#include <boost/thread.hpp>
#include "exer05-util.hpp"
using namespace std;
using namespace boost::assign;



typedef std::vector<double> vectord;
typedef std::vector<vectord> matrix;



void getProduct(const matrix& mat, const vectord& vec, vectord& result) {
    // Assume that size of mat and vec are both eligible
    int sizeRowMat = mat.size();
    int sizeColMat = mat[0].size();
    int sizeVec = vec.size();

    result.clear();
    result.resize(sizeRowMat, 0);

    boost::thread_group lstTh;

    for (int i = 0; i < sizeRowMat; ++i) {
        lstTh.add_thread(new boost::thread(
            &getScalarProduct, mat[i].data(), vec.data(), sizeVec, &result[i]
        ));
    }

    lstTh.join_all();
}



int main() {
    matrix A;

    {
        vectord row1, row2, row3;
        row1 += 1, 2, 3;
        row2 += 4, 5, 6;
        row3 += 7, 8, 9;
        A += row1, row2, row3;
    }

    vectord b;
    b += 3, -1, 0;

    vectord result;
    getProduct(A, b, result);

    for (int i = 0; i < result.size(); ++i) {
        cout << result[i] << endl;
    }

    return 0;
}
